net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,395 @@
+			
+	your_queue = q->prio_array[your_prio];
+
+	gk_len = calc_alloc_len(q);
+	if (gk_len < q->max_limit) {
+		err = fifo_set_limit(your_queue, 
+				    your_queue->limit + q->init_limit);
+		if (err) 
+			return err;
+	
+		q->init_limit += 1; /* Fast forward resource allocation*/
+	
+		ret = qdisc_enqueue(skb, your_queue, to_free);
+		if (ret != NET_XMIT_SUCCESS) {
+			if (net_xmit_drop_count(ret))
+				qdisc_qstats_drop(sch);
+		return ret;
+		}
+
+		qdisc_qstats_backlog_inc(sch, skb);
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+
+	} else if (your_prio == (0 || 1) && your_queue->limit > 0) {
+	
+		qdisc_drop(skb, sch, to_free);	
+	
+		/* Priority 0 queue is the default case queue and is not allowed 		 * to borrow or be borrowed from. It is treated as a best effort 		 * queue. Priority 1 queue is the least priority and so can't		         * borrow from anyone.  
+		 */
+	
+		qdisc_qstats_backlog_inc(sch, skb);
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+
+	} else {
+
+	/* Queues need to borrow resources from other queues */	
+		
+	if (your_queue->q.qlen == your_queue->limit && your_queue->limit > 0) {
+
+	prio = 1;
+
+	while (q->prio_array[prio]->limit == 0)
+		prio++;
+
+	if((q->prio_array[prio])->q.qlen > 0) {
+		__qdisc_queue_drop_head(q->prio_array[prio], 
+					&(q->prio_array[prio])->q, to_free);	
+		qdisc_qstats_drop(sch);
+	}
+
+	err = fifo_set_limit(q->prio_array[prio], 
+			    (q->prio_array[prio])->limit - 1);
+	if (err)
+		return err;
+	
+
+	err = fifo_set_limit(your_queue, your_queue->limit + 1);
+	if (err) 
+		return err;
+	
+	ret = qdisc_enqueue(skb, your_queue, to_free);
+	if (ret != NET_XMIT_SUCCESS) {
+		if (net_xmit_drop_count(ret))
+			qdisc_qstats_drop(sch);
+		return ret;
+	}
+
+	qdisc_qstats_backlog_inc(sch, skb);
+	sch->q.qlen++;
+	return NET_XMIT_SUCCESS;
+
+        }
+
+	if (your_queue->limit == 0) {
+		if (your_prio == 1 ||  
+		   (q->prio_array[your_prio + 1])->limit == 0) 
+			goto done;
+	}
+
+	/* Required priority queue is completey borrowed from and is not able 
+	 * to borrow either. No choice but to drop packet now.
+	 */
+
+	ret = qdisc_enqueue(skb, your_queue, to_free);
+	if (ret != NET_XMIT_SUCCESS) {
+		if (net_xmit_drop_count(ret))
+			qdisc_qstats_drop(sch);
+		return ret;
+	}
+	
+	qdisc_qstats_backlog_inc(sch, skb);
+	sch->q.qlen++;
+	return NET_XMIT_SUCCESS;	
+
+	}
+
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
+done:
+	q->last_dropped_prio = your_prio;
+	ret = qdisc_drop(skb, sch, to_free);
+	return ret;
+
+}
+
+static struct sk_buff *gkprioq_dequeue(struct Qdisc *sch)
+{
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	struct sk_buff *skb;
+	int err;
+	
+
+	skb = qdisc_dequeue_head(q->prio_array[q->highest_prio]);
+
+	if (skb) {
+		if (q->last_dropped_prio != -1) {
+			err = fifo_set_limit(q->prio_array[q->highest_prio], 
+				   (q->prio_array[q->highest_prio])->limit - 1);
```

Comment:
The problem with this approach is that you're trying to share resources from the queue with high priority (of the packet just dequeued) with the queue with low priority (of the packet just dropped). We don't want to give any extra resources to packets with low priority than is needed.

So you should definitely remove this functionality -- you should just enforce a global limit on the number of requests. If there's room under this limit for more requests, accept it. If there's not, just drop the lowest.